src/ray/object_manager/object_buffer_pool.h

```diff
@@ -0,0 +1,186 @@
+ private:
+  /// Seal the object associated with a create operation.
+  ray::Status SealCreate(const ObjectID &object_id);
+
+  /// Abort the create operation associated with an object.
+  ray::Status AbortCreate(const ObjectID &object_id);
+
+  /// Abort the get operation associated with an object.
+  ray::Status AbortGet(const ObjectID &object_id);
+
+  /// Builds the chunk vector for an object, and store it by object_id in chunk_info_.
+  /// Returns the number of chunks into which object is split.
+  uint64_t BuildChunks(const ObjectID &object_id, uint8_t *data, uint64_t data_size,
+                       uint64_t metadata_size);
+
+  /// Get an object from the object store pool.
+  std::shared_ptr<plasma::PlasmaClient> GetObjectStore();
+
+  /// Release an object from the object store pool.
+  void ReleaseObjectStore(std::shared_ptr<plasma::PlasmaClient> client);
+
+  /// Adds a client to the client pool and mark it as available.
+  void Add();
+
+  /// The type of a buffer state.
+  /// Used to detect whether an object buffer is
+  /// both a get and create buffer.
+  enum class BufferStateType : int { GET = 0, CREATE };
+
+  /// Holds the state of a buffer.
+  struct BufferState {
+    BufferState() {}
+    BufferState(std::shared_ptr<plasma::PlasmaClient> client, uint64_t references,
+                BufferStateType state_type)
+        : client(client), references(references), state_type(state_type) {}
+    /// The plasma client used by this buffer.
+    std::shared_ptr<plasma::PlasmaClient> client;
+    /// The number of references that currently rely on this buffer.
+    /// We expect this many calls to Release or SealOrAbortBuffer.
+    uint64_t references;
+    /// The type of buffer; either GET or CREATE.
+    BufferStateType state_type;
+    /// Whether a single failure has occurred.
+    bool one_failed = false;
+  };
+
+  /// Returned when Get fails.
+  ChunkInfo errored_chunk_ = {
+      nullptr, 0, 0, 0,
+      ray::Status::IOError("Unable to obtain object chunk, object not local.")};
+
+  /// Mutex for thread-safe operations.
+  std::mutex pool_mutex_;
+  /// Determines the maximum chunk size to be transferred by a single thread.
+  uint64_t chunk_size_;
+  /// A vector for each object maintaining information about the chunks which comprise
+  /// an object. ChunkInfo is used to transfer
+  std::unordered_map<ray::ObjectID, std::vector<ChunkInfo>, ray::UniqueIDHasher>
+      chunk_info_;
+  /// The state of a buffer that's currently being used.
+  std::unordered_map<ray::ObjectID, BufferState, ray::UniqueIDHasher> buffer_state_;
+  /// A set of vectors into which data is read for objects that failed to be created.
+  std::unordered_map<ray::ObjectID, std::vector<uint8_t>, ray::UniqueIDHasher>
+      create_failure_buffers_;
+  /// Tracks the number of calls to GetChunk that should fail after the first call
+  /// fails to obtain the buffer.
+  std::unordered_map<ray::ObjectID, uint64_t, ray::UniqueIDHasher> get_failed_count_;
```

Comment:
I don't really understand why this is necessary. Can we just have each of the `GetChunk` calls fail/succeed independently of each other?